packages/pin-input/src/use-pin-input.ts

```diff
@@ -156,166 +170,173 @@ export function usePinInput(props: UsePinInputProps = {}) {
     [values, setValues, focusNext, onComplete, descendants.length],
   )
 
-  const clear = useCallback(() => {
+  const clear = React.useCallback(() => {
     const values: string[] = Array(descendants.length).fill("")
     setValues(values)
     const firstInput = descendants[0]
     firstInput.element?.focus()
   }, [descendants, setValues])
 
-  return {
-    id,
-    domContext,
-    setValue,
-    values,
-    setValues,
-    setMoveFocus,
-    clear,
-    onComplete,
-    placeholder,
-    manageFocus,
-    isDisabled,
-    isInvalid,
-  }
-}
+  const getNextValue = React.useCallback(
+    (value: string, eventValue: string) => {
+      let nextValue = eventValue
+      if (value?.length > 0) {
+        if (value[0] === eventValue.charAt(0)) {
+          nextValue = eventValue.charAt(1)
+        } else if (value[0] === eventValue.charAt(1)) {
+          nextValue = eventValue.charAt(0)
+        }
+      }
+      return nextValue
+    },
+    [],
+  )
 
-export type UsePinInputReturn = ReturnType<typeof usePinInput>
+  const [focusedIndex, setFocusedIndex] = React.useState(-1)
 
-export interface UsePinInputFieldProps {
-  ref?: Ref<HTMLInputElement>
-  onChange?: ChangeEventHandler
-  onKeyDown?: KeyboardEventHandler
-  onFocus?: FocusEventHandler
-  onBlur?: FocusEventHandler
-}
+  const getInputProps = React.useCallback(
+    (props: InputProps & { index: number }): InputProps => {
+      const { index, ...rest } = props
 
-export function usePinInputField(props: UsePinInputFieldProps = {}) {
-  const { ref: forwardedRef, ...rest } = props
+      /**
+       * Improved from: https://github.com/uber/baseweb/blob/master/src/pin-code/pin-code.js
+       */
+      const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        const eventValue = event.target.value
+        const currentValue = values[index]
+        const nextValue = getNextValue(currentValue, eventValue)
 
-  const ref = useRef<HTMLInputElement>(null)
+        // if the value was removed using backspace
+        if (nextValue === "") {
+          setValue("", index)
+          return
+        }
 
-  const {
-    id,
-    isDisabled,
-    isInvalid,
-    setValue,
-    values,
-    setMoveFocus,
-    setValues,
-    domContext,
-    placeholder,
-    manageFocus,
-  } = usePinInputContext()
+        // in the case of an autocomplete or copy and paste
+        if (eventValue.length > 2) {
+          // see if we can use the string to fill out our values
+          if (validate(eventValue, type)) {
+            // Ensure the value matches the number of inputs
+            const nextValue = eventValue
+              .split("")
+              .filter((_, index) => index < descendants.length)
+
+            setValues(nextValue)
+
+            // if pasting fills the entire input fields, trigger `onComplete`
+            if (nextValue.length === descendants.length) {
+              onComplete?.(nextValue.join(""))
+            }
+          } else {
+            return
```

Comment:
I'm not sure what the UX here would look like, but should we potentially do something in this case or provide a callback to react to this case? Could be confusing to users to have their paste just not work.

Reply:
We need to be a bit strict here given the user can enter both "alphanumeric" or "number" values. If developers want users to enter any value, then it's appropriate to use "alphanumeric".

We also want to be able to sanitize/validate the values user paste into the box (out of security concerns) with what the business logic allows.

You get what I mean?